---
title: "Genome Assembly"
author: "Prof. Harbert"
date: Meeting 16
output: 
  html_document:
    theme: united
    highlight: tango
    df_print: paged
    toc: true
    toc_float: true
    toc_depth: 4
  
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Genome Assembly

Challenge: Assemble a 1 billion piece puzzle where the pieces are all made of only 4 colors and repeated shapes.

```{r echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics('./images/genome_assembly.png')
```

## Setup

In late December of 2019 doctors and scientists in Wuhan, China were tracking down the cause of an unknown respiritory virus. One part of this effort was to sequence RNA from bronchealveolar fluid from one of the patients. Those data are now available for us to work with:

[Data from 9 samples sequenced on December 30, 2019](https://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=605983)

[Publication of the resulting genome](https://doi.org/10.1038/s41586-020-2012-7)

All of these samples are raw RNA sequencing data (represented here in the DNA alphabet) that were generated on an Illumina platform instrument resulting in paired end reads of 250 bp in length. This is a common, but tough, assembly scenario.


```{bash, eval = FALSE}
mkdir data

#may try also https://sra-pub-sars-cov2.s3.amazonaws.com/run/SRR11092059/SRR11092059
#fasterq-dump -e 4 --split-files -O data $SRA

wget -bqc --output-document data/SRR11092061.R1.fq.gz https://sra-pub-sars-cov2.s3.amazonaws.com/sra-src/SRR11092061/v300043428_L04_123_1.fq.gz 
wget -bqc --output-document data/SRR11092061.R2.fq.gz https://sra-pub-sars-cov2.s3.amazonaws.com/sra-src/SRR11092061/v300043428_L04_123_1.fq.gz 

wget -bqc --output-document data/SRR11092058.R1.fq.gz https://sra-pub-sars-cov2.s3.amazonaws.com/sra-src/SRR11092058/ICU4G_S1_L001_R1_001.fastq.gz 
wget -bqc --output-document data/SRR11092058.R2.fq.gz https://sra-pub-sars-cov2.s3.amazonaws.com/sra-src/SRR11092058/ICU4G_S1_L001_R2_001.fastq.gz

wget -bqc --output-document data/SRR11092060.R1.fq.gz https://sra-pub-sars-cov2.s3.amazonaws.com/sra-src/SRR11092060/v300043428_L04_124_1.fq.gz
wget -bqc --output-document data/SRR11092060.R2.fq.gz https://sra-pub-sars-cov2.s3.amazonaws.com/sra-src/SRR11092060/v300043428_L04_124_2.fq.gz

# unzip all of the data files using gunzip
ls data/*.fq.gz | xargs -P 6 -n 1 gunzip {}



```
## QC

Evaluate the read qualities for this dataset using fastqc. Are there any red flags?

```{bash, eval=F}
SRA=SRR11092061
fastqc -t 2 data/$SRA*.fq
```


## Muddy data: Clean it up

The data in these samples is bulk RNA. Therefore it likely contains a lot of mRNA transcripts from the human host and any bacteria present in the sample. *Some* of these reads might come from the causative viral agent if we are lucky.

One thing we can do to help get us on the right track for assembly is to align these reads to other viral genomes to filter for reads that at least *look like* they came from a virus genome.

```{bash, eval=F}
wget ftp://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.1.1.genomic.fna.gz
wget ftp://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.2.1.genomic.fna.gz
wget ftp://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.3.1.genomic.fna.gz
mv viral.*.fna.gz data/
gunzip data/viral.*.fna.gz
cat data/viral.*.fna > data/virus_all.fna

```

Index genome and align reads to all known virus genomes. Then convert SAM back to fastq for *only those reads that map to any virus genome*.

```{bash, eval=F}

bwa index data/virus_all.fna

#for infile in data/*.R1.fq
#do
#  base=$(basename ${infile} .R1.fq)
#  bwa mem -t 8 data/virus_all.fna data/${base}.R1.fq data/${base}.R2.fq > data/${base}.sam
#done 

#or cat all the R1 and R2 files and align/filter together
cat data/*.R1.fq > data/uknownRNA.R1.fq
cat data/*.R2.fq > data/uknownRNA.R2.fq
bwa mem -t 8 data/virus_all.fna data/uknownRNA.R1.fq data/uknownRNA.R1.fq > data/unknown.sam

samtools view -@ 8 -S -b data/unknown.sam > data/unknown.align.bam
samtools view -@ 8 -b -F4 data/unknown.align.bam > data/unknown.align.final.bam
samtools sort -@ 8 data/unknown.align.final.bam > data/unknown.align.sort.bam
samtools index -@ 8 data/unknown.align.sort.bam
samtools fastq -@ 8 data/unknown.align.sort.bam -1 data/unknown.filter.R1.fastq -2 data/unknown.filter.R2.fastq -s data/unknown.filter.single.fastq



### old version
#bwa mem -t 8 data/virus_all.fna data/$SRA\_1.fastq data/$SRA\_2.fastq > data/$SRA.sam

#samtools view -@ 8 -S -b data/$SRA.sam > data/$SRA.align.bam
#samtools view -@ 8 -b -F4 data/$SRA.align.bam > data/$SRA.align.final.bam
#samtools sort -@ 8 data/$SRA.align.final.bam > data/$SRA.align.sort.bam
#samtools index -@ 8 data/$SRA.align.sort.bam
#samtools fastq -@ 8 data/$SRA.align.sort.bam -1 data/$SRA.filter.R1.fastq -2 data/$SRA.filter.R2.fastq -s data/$SRA.filter.single.fastq


```


## Programs for De Novo Assembly

[Illumina technical note on short read de novo assembly](https://www.illumina.com/Documents/products/technotes/technote_denovo_assembly_ecoli.pdf)

[Melbourne Bioinformatics De Novo Assembly tutorial](https://www.melbournebioinformatics.org.au/tutorials/tutorials/assembly/assembly-background/)

[What is N50](http://www.molecularecologist.com/2017/03/whats-n50/)

All assembly programs work out from related core principles: In short they all look for short overlapping pieces of every read in a dataset and build a graph of these connections. This structure can then be traversed to find the longest pieces of contiguous overlap across many reads. Longer contigs can then collapse parts of the graph until there are no further overlaps.

The data structure involved here is known as a de Bruijn graph:

```{R echo=FALSE, out.width='100%', fig.align='center'}
download.file('https://cameroncounts.files.wordpress.com/2015/02/db.jpg?w=453&zoom=2', 'images/debruijn.jpg')
knitr::include_graphics('images/debruijn.jpg')

```

We will try three such programs on our raw data here to see if we can find contiguous sequence in these bronchealveolar fluid RNA sequencing samples.

### Velvet: A classic 

Using velvet consists of a sequence of two commands:

+ velveth - analyzes kmers in the reads
+ velvetg - constructs and refines the assembly graph

Run the following and take a coffee break:

```{bash, eval=F}
#make an output directory
mkdir dirAssembly

#set environmental parameter that controls velvet parallel processing
# will run *parts* of the next steps up to 8x faster than single core operation
export OMP_NUM_THREADS=8

# run velveth: kmers
velveth dirAssembly 31 -shortPaired -fastq -separate data/unknown.filter.R1.fastq data/unknown.filter.R2.fastq

#run velvetg: assembly 
velvetg dirAssembly 

```

Look at the contigs file:

```{R, eval=T}
library(Biostrings)
library(ggplot2)
contigs = readDNAStringSet('dirAssembly/contigs.fa', format='fasta')
print(contigs)
maxcontig=max(contigs@ranges@width)
maxcontigSTR = contigs[which(contigs@ranges@width == maxcontig)]
as.character(maxcontigSTR)

widths = data.frame(contigs@ranges@width)
names(widths) = 'contig_length'
ggplot(widths) + 
  geom_density(aes(x=contig_length))

```

Check the sequence of our longest contig with web BLAST: https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome


### SPAdes

```{bash, eval=F}

```

### MEGAHIT: Specialized for metagenome (mixtures) assembly

```{bash, eval=F}

wget https://github.com/voutcn/megahit/releases/download/v1.2.9/MEGAHIT-1.2.9-Linux-x86_64-static.tar.gz
tar zvxf MEGAHIT-1.2.9-Linux-x86_64-static.tar.gz
mv MEGAHIT-1.2.9-Linux-x86_64-static MEGAHIT
./MEGAHIT/bin/megahit -m 0.5 -t 8 -1 data/$SRA\_1.fastq -2 data/$SRA\_2.fastq -o dirMEGAHIT
./MEGAHIT/bin/megahit -m 0.5 -t 8 -1 data/unknown.filter.R1.fastq -2 data/unknown.filter.R2.fastq -o dirMEGAHIT

```

Let's look again at the contigs file:

```{R, eval=T}
library(Biostrings)
library(ggplot2)
contigs = readDNAStringSet('dirMEGAHIT/final.contigs.fa', format='fasta')
print(contigs)
maxcontig=max(contigs@ranges@width)
maxcontigSTR = contigs[which(contigs@ranges@width == maxcontig)]
as.character(maxcontigSTR[[1]])


```




# Homework: 

[Read the Illumina technical note on genome assembly](https://www.illumina.com/Documents/products/technotes/technote_denovo_assembly_ecoli.pdf) and post an explanation of one figure or table to slack #discussion.

[home](https://bio331.devbioinformatics.org)