<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>DNA Sequencing</title>
    <meta charset="utf-8" />
    <meta name="author" content="Prof. Harbert" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="zh-CN.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# DNA Sequencing
## BIO331 Introduction to Bioinformatics
### Prof. Harbert

---


## DNA Sequencing: Overview

The ability to sequence DNA is one of the greatest achievements of the field of Biology. 

We will talk today a bit about the history of DNA sequencing and the main technologies involved.

Topics:

Traditional
+ Sanger Sequencing 
+ The Human Genome Project

Modern Sequencing Technologies
+ Illumina
+ PacBio
+ NanoPore
+ BGI

Collaborative Research (HW)

---



## Sanger

Dr. Frederick Sanger (1918-2013) -- Developed sequencing by the chain terminating dideoxynucleotide method that now carries his name.
[
&lt;img src="./images/Frederick_Sanger2.jpg" width="30%" style="display: block; margin: auto;" /&gt;
](https://en.wikipedia.org/wiki/Frederick_Sanger#/media/File:Frederick_Sanger2.jpg)

---


## DNA Polymerase

Nucleic acid molecules are copied in living cells through the enzymatic activity of nuclease enzymes.

Polymers of nucleotides are formed by covalent bonds between the phosphate and sugar groups.

New nucleotides are always added on the 3' sugar carbon.


&lt;img src="./images/dna_structure.jpg" width="50%" style="display: block; margin: auto;" /&gt;

[OpenStax College, Biology (CC BY 4.0)](https://openstax.org/books/concepts-biology/pages/9-1-the-structure-of-dna#fig-ch09_01_03)

---


## Sanger

&lt;img src="./images/fig1.jpg" width="50%" style="display: block; margin: auto;" /&gt;

[Image credit: "Whole-genome sequencing: Figure 1," by OpenStax College, Biology (CC BY 4.0)](https://cnx.org/contents/GFy_h8cu@10.7:5l844Z38@7/Whole-Genome-Sequencing)

---



## Sanger

&lt;img src="./images/fig2.jpg" width="50%" style="display: block; margin: auto;" /&gt;

[Image credit: "Whole-genome sequencing: Figure 1," by OpenStax College, Biology (CC BY 4.0)](https://cnx.org/contents/GFy_h8cu@10.7:5l844Z38@7/Whole-Genome-Sequencing)

---


## Sanger

&lt;div align="center"&gt;
&lt;iframe width="560" height="315" src="https://www.youtube.com/embed/FvHRio1yyhQ?rel=0"; frameborder="0" allowfullscreen&gt;&lt;/iframe&gt;
&lt;/div&gt;

---


## Sanger

&lt;div align="center"&gt;
&lt;iframe width="560" height="315" src="https://www.youtube.com/embed/e2G5zx-OJIw?rel=0"; frameborder="0" allowfullscreen&gt;&lt;/iframe&gt;
&lt;/div&gt;

---


## Sanger

Produces high quality reads
+ Length: 400-600 bp (sometimes longer)
+ High base call accuracy
+ Maxes out at about 384 reads per run on typical equipment (some of this can be automated).
+ A few hours per run

How many cycles at 500bp per read and (let's assume) 384 reads per run would it take to sequence the human genome at 3 billion bases one time?

---


## Human Genome Project


&lt;img src="./images/hgp_timeline.jpg" width="100%" style="display: block; margin: auto;" /&gt;

[Source CC-by-2.0](https://www.flickr.com/photos/genomegov/26964377742/)

[It was a BIG deal!](https://www.youtube.com/watch?v=slRyGLmt3qc)

---

## Massively Parallel Sequencing

+ Illumina
+ PacBio
+ Nanopore
+ BGI

---


## Illumina

&lt;div align="center"&gt;
&lt;iframe width="560" height="315" src="https://www.youtube.com/embed/fCd6B5HRaZ8?rel=0"; frameborder="0" allowfullscreen&gt;&lt;/iframe&gt;
&lt;/div&gt;

---


## PacBio

&lt;div align="center"&gt;
&lt;iframe width="560" height="315" src="https://www.youtube.com/embed/TwN9EaDtgjU?rel=0"; frameborder="0" allowfullscreen&gt;&lt;/iframe&gt;
&lt;/div&gt;

---


## PacBio
&lt;div align="center"&gt;
&lt;iframe width="560" height="315" src="https://www.youtube.com/embed/NHCJ8PtYCFc?rel=0"; frameborder="0" allowfullscreen&gt;&lt;/iframe&gt;
&lt;/div&gt;

---


## Ion Torrent

&lt;div align="center"&gt;
&lt;iframe width="560" height="315" src="https://www.youtube.com/embed/WYBzbxIfuKs?rel=0"; frameborder="0" allowfullscreen&gt;&lt;/iframe&gt;
&lt;/div&gt;

---


## Oxford Nanopore

&lt;div align="center"&gt;
&lt;iframe width="560" height="315" src="https://www.youtube.com/embed/E9-Rm5AoZGw?rel=0"; frameborder="0" allowfullscreen&gt;&lt;/iframe&gt;
&lt;/div&gt;

---


## Other Resources

Library prep and Illumina Seq: http://rmpiro.net/teaching/pub/lectures/fu-genomics/01-NGS_technology.pdf

Review of sequencing platforms (2017):
https://www.omicsonline.org/open-access/generations-of-sequencing-technologies-from-first-to-next-generation-0974-8369-1000395.php?aid=87862



---


## Homework

Research one sequencing technology and contribute to the shared class file.

[Home](https://bio331.devbioinformatics.org)

---
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
