---
title: "Metagenomics 2"
author: "Prof. Harbert"
date: Meeting 20
output: 
  html_document:
    theme: united
    highlight: tango
    df_print: paged
    toc: true
    toc_float: true
    toc_depth: 4
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Fast(er) Altorithms

There are many methodological approaches to classify large amounts of DNA sequence data in more efficient ways that maintain or improve upon the accuracy of BLAST. In general these methods involve generating complex data structures from the reference data that are relatively efficient to search.

## Kraken(2)

[Wood, D.E. and Salzberg, S.L., 2014. Kraken: ultrafast metagenomic sequence classification using exact alignments. Genome biology, 15(3), p.R46.](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-3-r46)

Uses a kmer classification algorithm that breaks the reference data into chunks of sequence (kmers). The kmers are then sorted into a tree-like data structure so that they can be easily and efficiently searched.

[Manual](https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown)


### Installation

```{bash, eval=F,}
cd ~/src
git clone https://github.com/DerrickWood/kraken2
cd kraken2
./install_kraken2.sh ./
```

Then add to ~/.bashrc

```{bash, eval=F}
export PATH=$PATH:~/src/kraken2
```

### Create a Kraken Database

*DO NOT RUN* Building the kraken database takes A LOT of time (*mostly getting the taxonomy files set up*). Your instructor will provide a path to the appropriate database files.

```{bash, eval=F}
kraken2-build --download-library "viral" --threads 8 --db krakendb
kraken2-build --download-taxonomy --threads 8 --db krakendb
kraken2-build --build --db krakendb

```
### Query with Kraken2

We will implement two different queries with kraken2: 1) using the default settings, and 2) using a higher confidence threshold (similar to our BLAST filtering strategy).

```{bash, eval=T}
kdb=krakendb
targdata=data/viral_metagenome.fq


#defaults
kraken2 --db $kdb  --threads 8 --use-names --report kreport.tab $targdata > kraken.out

# high confidence (not a probability scoring system -- confidence values are arbitrary)
kraken2 --db $kdb  --threads 8 --confidence 0.9 --use-names --report kreport_conf.tab $targdata > kraken_conf.out

```


# Visualizing Kraken report files

Use ggplot to show the # of reads mapping to genera.

```{R}
library(ggplot2)

kreport=read.delim('kreport.tab', sep ='\t', header=F)

#preview
head(kreport)

#how many rows?
nrow(kreport)

#add column names to the kreport table
colnames(kreport) = c('percent', 'readsRooted', 'reads', 'taxRank', 'taxID', 'sciName')

ggplot(data=kreport[which(kreport$taxRank=='F'),]) +
  geom_bar(aes(x=sciName, y = readsRooted), stat='identity') +
  theme(axis.text.x = element_text(angle=90))

ggplot(data=kreport[which(kreport$taxRank=='S'),]) +
  geom_bar(aes(x=sciName, y = readsRooted), stat='identity') +
  theme(axis.text.x = element_text(angle=90))

```

```{R}
kreport_conf=read.delim('kreport_conf.tab', sep ='\t', header=F)

#add column names to the kreport table
colnames(kreport_conf) = c('percent', 'readsRooted', 'reads', 'taxRank', 'taxID', 'sciName')

ggplot(data=kreport_conf[which(kreport_conf$taxRank=='F'),]) +
  geom_bar(aes(x=sciName, y = readsRooted), stat='identity') +
  theme(axis.text.x = element_text(angle=90))

ggplot(data=kreport_conf[which(kreport_conf$taxRank=='S'),]) +
  geom_bar(aes(x=sciName, y = readsRooted), stat='identity') +
  theme(axis.text.x = element_text(angle=90))

```