---
title: "rBLAST"
author: "Prof. Harbert"
output: 
  html_document:
    theme: united
    highlight: tango
    df_print: paged
    toc: true
    toc_float: true
    toc_depth: 4
    
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# BLAST: Basic Local Alignment Search Tool

BLAST (Basic Local Alignment Search Tool) optimizes time complexity in pairwise alignment by searching for short, local, sequence matches and then extending the comparison out from the matching regions.

Some good resources to review later:

+ http://www.metagenomics.wiki/tools/blast
+ https://www.ndsu.edu/pubweb/~mcclean/plsc411/Blast-explanation-lecture-and-overhead.pdf

## Web BLAST

Most people first interact with BLAST via the [web portal](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome).

In the web portal search for this sequence:

"CATTTCAGGTGTGGGTGTTTATGATTGCGCCCTACCGTGACGTTTTCATCTATCGGAGGGAGTGGAACACCTGTCGCTTTCTATCTTCCATTGTACTTTGTGTTTACGAGCTAAAGTTTTAACACAAGAAGTTAGTATATTTACCATTCGATACAAACTCTTTTTTTTTGAGGACCGCTGTAAATAATTTCTGCATATGCATTACGGCGATCGTGATAATAGAAAAATCCGTCGGTCAGTCGACTTACTAACAAGGATGTCCTAGTACGCTTCACAAAATTTCATTTTTGTAATGATCCAATCCAGAGACTGATTAGACTAATGTAACATCAGCTTTCTTAATAGCATTATCCGTTATAAATGAATTTTCGGCGTTTGACTCCGCTTCAAATATTTTGTCGCACACTTGAAAGATAGCCAAAGCCGAAGAATGCTTGGTTGATTGTATAGATCCTTCCGGGACTGAGCCCACCCATAAAATGCCATTGCCATAAACGAACAAGGTAACGGCCTGTTTCCATTTATTCATTCAGAAGTATCTTTTGTAGCCTTCCAAAATGGATTTTCCCCGATATCTAACATAATGCATGAAAGGATCTTTGAAGACCCGTCTGCGATAGCCGGAAAATCATTAGCAAAGACTTCTTCTACCACAAGGATATTTTTATTTTCCATAGAAATATATTCAGCTCAAAAAAAAGCCCCATGAGAGGATGCGACCCAATCGTAAATGAGAAAAGTGGTTGCGGAGAAAAAGTAAGATGGATTCGTATTCTAAGCATGAGAATTATAGGAACAAGGAAGCCTAATCTTGGATTTATTTTGCAGAAAAGGAAAAGTCAAATAGATTTTTGTAGAAAATAATAAAAACATTACCTTCAATTAGAATATAGATGAAGAAAAGGCCGTAATAAATGTACAAAGAAGGAATGCCACTTGTGAAGAATTTGAACCAAGATTTCCAGATGGACTGGGTGCAATATCAGCACCAACAGGAAATCCAGCGTTTCCTCGATAGATGAAACGT"

What is it? ^^^

How can we tell?

How are E values like p-values?

*We will revisit command line BLAST tools in our later metagenomics unit*

# rBlast

[The rBLAST package](https://github.com/mhahsler/rBLAST/) provides R code tools that implement access to the commandline BLAST tools in R. Today we will use this package to build a new BLAST database, run BLAST queries, and visualize the results in ggplot.

## Getting Started

```{R, message=F}
#install.packages('devtools')
#devtools::install_github("mhahsler/rBLAST")
library(rBLAST)
library(ggplot2)
```

## Get data

Download FASTQ sequence read data from the NCBI SRA:

```{bash, eval=F}
fastq-dump SRR13764788
```

## Read into R

```{R}
fq <- readDNAStringSet('SRR13764788.fastq',format = 'fastq')
head(fq)
```

*Think: What tools can we use to look at these FASTQ sequence reads?*

## Database: Reference Genome

```{R}
download.file('ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/Escherichia_coli/reference/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz',
              'genome.fna.gz')
system('gunzip genome.fna.gz')
genome = readDNAStringSet('genome.fna')
print(genome)
```

*Think: What tools do we have to explore this genome?*


## Setting up BLAST in R

Set variables to reference (genome) and reads (SRA reads file).

```{r}
reference = 'genome.fna'
reads = 'SRR13764788.fastq'
dna <- readDNAStringSet(reads, format='fastq')
geno.freq = as.data.frame(alphabetFrequency(dna))
dna <- dna[geno.freq$N==0] 
```


Make BLAST database from reference.

```{R}
#makeblastdb creates a folder that contains a blast database 
makeblastdb(reference, dbtype = "nucl")
```

Prepare a BLAST query

```{R, message=F}
#prepare
bl <- blast(db=reference, type='blastn')
#run
cl <- predict(bl, dna, BLAST_args="-max_target_seqs 1")

cl[1:5,]
#to view first 5 hits

```
## Visualize BLAST hits

```{R}
ggplot(cl) +
  geom_histogram(aes(x=Perc.Ident)) +
  theme_minimal()

ggplot(cl) +
  geom_density(aes(x=S.start), 
               kernel = 'rectangular', 
               n=4600,
               col='darkorchid') +
  theme_minimal() +
  xlab("Genomic Position") +
  ylab("Density of BLAST Hits")

```



[home](https://bio331.devbioinformatics.org)

